pep-0999.rst

```diff
@@ -0,0 +1,541 @@
+       directors: NotRequired[List[str]]
+       writers: NotRequired[List[str]]
+
+It is an error to use ``Required[...]`` on an item of a TypedDict whose
+totality is true:
+
+::
+
+   class Movie(TypedDict):   # implicitly total=True
+       title: Required[str]  # error: Required[...] is only usable when total=False
+
+::
+
+   class Movie(TypedDict, total=True):
+       title: Required[str]  # error: Required[...] is only usable when total=False
+
+It is an error to use ``NotRequired[...]`` on an item of a TypedDict
+whose totality is false:
+
+::
+
+   class Cell(TypedDict, total=False):
+       value: NotRequired[object]  # error: NotRequired[...] needs total=True
+
+It is an error to use ``Required[...]`` or ``NotRequired[...]`` in any
+location that is not an item of a TypedDict:
+
+::
+
+   # error: Required[...] is only usable on items of TypedDict
+   password: Required[str]
+
+   # error: Required[...] is only usable on items of TypedDict
+   def login(password: Required[str]) -> None: ...
+
+   # error: Required[...] is only usable on items of TypedDict
+   def get_signature() -> Required[Signature]: ...
+
+
+Backwards Compatibility
+=======================
+
+No backward incompatible changes are made by this PEP.
+
+
+How to Teach This
+=================
+
+To define a TypedDict where most keys are required and some are
+potentially-missing, you should define a single TypedDict and mark those
+few keys that are potentially-missing by wrapping the value type with
+``NotRequired[...]``. For example:
+
+::
+
+   class Movie(TypedDict):
+       title: str
+       release_year: int
+       directors: NotRequired[List[str]]  # mark potentially-missing key
+       writers: NotRequired[List[str]]    # mark potentially-missing key
+
+If on the other hand most keys are potentially-missing and a few are
+required, you should instead define a single TypedDict with
+``total=False`` and mark those few keys that are required by wrapping
+the value type with ``Required[...]``. For example:
+
+::
+
+   class Node(TypedDict, total=False):
+       value: Required[object]            # mark required key
+       label: str
+       weight: float
+
+If some items accept ``None`` in addition to a regular value, it is
+recommended that the ``TYPE|None`` syntax be preferred over
+``Optional[TYPE]`` for marking such item values, to avoid using
+``Required[...]`` or ``NotRequired[...]`` alongside ``Optional[...]``
+within the same TypedDict definition:
+
+Yes:
+
+::
+
+   from __future__ import annotations  # for Python 3.7-3.9
+
+   class Dog(TypedDict):
+       name: str
+       owner: NotRequired[str|None]
+
+Avoid (unless Python 3.5-3.6):
+
+::
+
+   class Dog(TypedDict):
+       name: str
+       # ick; avoid using both Optional and NotRequired
+       owner: NotRequired[Optional[str]]
+
+
+Reference Implementation
+========================
+
+The following will be true when
+`mypy#9867 <https://github.com/python/mypy/issues/9867>`__ is
+implemented:
```

Comment:
That issue is closed. It's not customary to link to speculative implementations from a PEP under "reference implementation." It's okay however to say something vague like "I'm working on it." (Once you have some draft code you can submit a PR that updates the PEP with the PR. You'll probably need at least three PRs, one for the stdlib, one for typing_extensions, and one for mypy.)

Reply:
Whoops. I used the wrong link. Will update.

Did intend to link to: https://github.com/python/mypy/issues/9773

Reply:
Nope that's the wrong PEP. Altered language again:

```
The goal is to be able to make the following statement:

    ...

The mypy implementation is currently still being worked on.
```